src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -21706,6 +21706,484 @@ void M(string? s)
+        {
+            var lib_cs =
+@"using System.Diagnostics.CodeAnalysis;
+public static class Extension
+{
+    [return: NotNullIfNotNull(""p"")] public static string? M(this string? p) => throw null!;
+}";
+            var source = @"
+class D
+{
+    static void M(string? p1, string? p2)
+    {
+        _ = p1 ?? throw null!;
+        p1.M().ToString();
+        p2.M().ToString(); // 1
+    }
+}";
+            var expected = new[]
+            {
+                // (8,9): warning CS8602: Dereference of a possibly null reference.
+                //         p2.M().ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "p2.M()").WithLocation(8, 9)
+            };
+
+            var lib = CreateNullableCompilation(new[] { NotNullIfNotNulltAttributeDefinition, lib_cs });
+            var comp = CreateNullableCompilation(source, references: new[] { lib.EmitToImageReference() });
+            comp.VerifyDiagnostics(expected);
+
+            var comp2 = CreateNullableCompilation(new[] { source, lib_cs, NotNullIfNotNulltAttributeDefinition });
+            comp2.VerifyDiagnostics(expected);
+            CompileAndVerify(comp2);
+        }
+
+        [Fact]
+        public void NotNullIfNotNull_OutParameter()
+        {
+            var lib_cs =
+@"using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    public void M(string? p, [NotNullIfNotNull(""p"")] out string? p2) => throw null!;
+}";
+            var source = @"
+class D
+{
+    static void M(C c, string? p1, string? p2)
+    {
+        _ = p1 ?? throw null!;
+        c.M(p1, out var x1);
+        x1.ToString();
+
+        c.M(p2, out var x2);
+        x2.ToString(); // 1
+    }
+}";
+            var expected = new[]
+            {
+                // (11,9): warning CS8602: Dereference of a possibly null reference.
+                //         x2.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x2").WithLocation(11, 9)
+            };
+
+            var lib = CreateNullableCompilation(new[] { NotNullIfNotNulltAttributeDefinition, lib_cs });
+            var comp = CreateNullableCompilation(source, references: new[] { lib.EmitToImageReference() });
+            comp.VerifyDiagnostics(expected);
+
+            var comp2 = CreateNullableCompilation(new[] { source, lib_cs, NotNullIfNotNulltAttributeDefinition });
+            comp2.VerifyDiagnostics(expected);
+            CompileAndVerify(comp2);
+        }
+
+        [Fact]
+        public void NotNullIfNotNull_OutParameter_WithMaybeNullWhen()
+        {
+            var lib_cs =
+@"using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    public bool M(string? p, [NotNullIfNotNull(""p""), MaybeNullWhen(true)] out string p2) => throw null!;
+}";
+            var source = @"
+class D
+{
+    static void M(C c, string? p1, string? p2)
+    {
+        _ = p1 ?? throw null!;
+        _ = c.M(p1, out var x1) 
+            ? x1.ToString()
```

Comment:
Do we have the precedence of these operators defined somewhere? For example, what if I had a `MaybeNull` and a `NotNullIfNotNull` on the same parameter?

Reply:
I think we should use the order in which the attributes are applied ;-) (just kidding)
I'll update the doc as possible

---
In reply to: [303563359](https://github.com/dotnet/roslyn/pull/37201#discussion_r303563359) [](ancestors = 303563359)

Reply:
The current order is NotNullIfNotNull > MaybeNull > NotNull.
Basically, NotNullIfNotNull wins over all other attributes (including conditional ones like MaybeNullWhen). Tested by  `NotNullIfNotNull_OutParameter_WithMaybeNullWhen` and `NotNullIfNotNull_Return_WithMaybeNull`.

---
In reply to: [303633923](https://github.com/dotnet/roslyn/pull/37201#discussion_r303633923) [](ancestors = 303633923,303563359)